src/components/pages/home.jsx

```diff
@@ -32,28 +34,32 @@ const HomePage = React.createClass({
     return {inputValue: ""}
   },
   handleClearSearchClicked() {
-    this.setState({inputValue: ""})
-    this.context.router.push("")
+    const query = ""
+    this.setState({inputValue: query})
+    this.context.setQuery(query)
+    this.context.router.push(query)
   },
   handleInputChange(event) {
     this.setState({inputValue: event.target.value})
     this.searchInput.scrollIntoView()
   },
   handleSubmit(event) {
     event.preventDefault()
+    this.context.setQuery(this.state.inputValue)
     this.context.router.push(`?q=${this.state.inputValue}#${searchInputId}`)
   },
-  locationHasChanged(location) {
-    if (this._isMounted) {
-      const query = location.query.q || ""
+  locationHasChanged(nextLocation) {
+    const {location} = this.props
+    // Check that the new location stays on the Home page, to avoid overwriting the query in App state.
+    if (this._isMounted && location.pathname === nextLocation.pathname) {
```

Comment:
Why not use [`isActive`](https://github.com/ReactTraining/react-router/blob/master/docs/API.md#isactivepathorloc-indexonly)?